File: 0684-redundant-connection/0684-redundant-connection.cpp
class Solution {
public:
    vector<int> parent;
    int find(int u) {
        return parent[u] == u ? u : parent[u] = find(parent[u]);
    }
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size();
        parent.resize(n + 1);
        for (int i = 1; i <= n; ++i)
            parent[i] = i;
        for (auto& e : edges) {
            int u = e[0], v = e[1];
            int pu = find(u), pv = find(v);
            if (pu == pv)
                return e;
            parent[pu] = pv;
        }
        return {};
    }
};

File: 0684-redundant-connection/0684-redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        def find(parent, u):
            if parent[u] != u:
                parent[u] = find(parent, parent[u])  
            return parent[u]
        
        def union(parent, u, v):
            pu = find(parent, u)
            pv = find(parent, v)
            if pu == pv:
                return False  
            parent[pu] = pv
            return True  
        
        n = len(edges)
        parent = [i for i in range(n + 1)] 
        
        for u, v in edges:
            if not union(parent, u, v):
                return [u, v] 
        return []